### ISSUE TITLE
YOLO dataset

### ISSUE THREAD

**Issue by jl749** [jl749/YOLOv3#2]
## pre-defined anchors (common obj ratio found via K-means)
```
anchors = tensor([[0.2800, 0.2200],  # pre defined
                  [0.3800, 0.4800],
                  [0.9000, 0.7800],
                  [0.0700, 0.1500],
                  [0.1500, 0.1100],
                  [0.1400, 0.2900],
                  [0.0200, 0.0300],
                  [0.0400, 0.0700],
                  [0.0800, 0.0600]])
```

## label txt file
```
# class, x, y, w, h
8 0.764 0.6069277108433735 0.23600000000000002 0.3042168674698795
8 0.594 0.6159638554216867 0.188 0.29819277108433734
14 0.229 0.6445783132530121 0.166 0.45180722891566266
14 0.39 0.6430722891566265 0.168 0.4307228915662651
14 0.5650000000000001 0.5918674698795181 0.154 0.41867469879518077
14 0.787 0.5963855421686747 0.166 0.3855421686746988
```
all x, y, w, h scales are within 0~1 range (normalized)

## VOCDataset
`VOCDataset: torch.utils.data.Dataset` overrides `__getitem__` method to adjust `label.txt` x, y, w, h values and return appropriate scale's cell-relative coordinates (e.g. [0.9320, 0.4223, 3.0680, 2.6239] is the relative coor to the first anchor box in scale 0)

**RETURN --> img:(C, W, H) && expected_bbox_info:( (3, 13, 13, 6), (3, 26, 26, 6), (3, 52, 52, 6) )**

```
loop expected bboxes (txt file):
    coor_from_txt = [0.764, 0.6069277108433735, 0.23600000000000002, 0.3042168674698795]
    IoU_wh(coor_from_txt[2:4],
                          anchors)  # calculate IoU with width and height
    IoU_arg_sorted = [0, 5, 1, 4, 3, 2, 8, 7, 6]  # coor_from_txt most likely to match the first anchor box in scale0
    anchor_indices = IoU_arg_sorted

    highest IoU anchor box ratio = (0.28, 0.22) <-- from index 0
    index 0 means --> anchor belongs to the first prediction (3, 13, 13, 6) && first anchor box out of three
    
    now RESCALE...
```
![image](https://user-images.githubusercontent.com/67103130/150452423-a8e587c8-fcb5-4d2d-9b89-baff2698d960.png)
there are 3 scales `S=(13, 26, 52)`
scale0 = first prediction `(3, 13, 13, 6)` 3 anchor boxes each (obj_prob, x, y, w, h, class)
scale1 = second prediction `(3, 26, 26, 6)`
scale3 = third prediction `(3, 52, 52, 6)`
https://github.com/jl749/YOLOv3/blob/a59f9f79ab558c766a160b0b0660b0724ce015f0/yolov3/datasets/pascal_VOC.py#L69-L99

**Comment by jl749** [comment]
https://github.com/jl749/YOLOv3/blob/852a785742dd6fbfad6d1d93775c119ae88c999a/yolov3/config.py#L34-L38

unnormalize anchors by multiplying scales (anchors --> scaled_anchors)
https://github.com/jl749/YOLOv3/blob/a59f9f79ab558c766a160b0b0660b0724ce015f0/yolov3/datasets/pascal_VOC.py#L121-L134

**Comment by jl749** [comment]
## torch.LongTensor vs torch.Tensor

Data type | dtype | CPU tensor | GPU tensor
-- | -- | -- | --
32-bit floating point | torch.float32 or torch.float | torch.FloatTensor | torch.cuda.FloatTensor
64-bit floating point | torch.float64 or torch.double | torch.DoubleTensor | torch.cuda.DoubleTensor
16-bit floating point | torch.float16 or torch.half | torch.HalfTensor | torch.cuda.HalfTensor
8-bit integer (unsigned) | torch.uint8 | torch.ByteTensor | torch.cuda.ByteTensor
8-bit integer (signed) | torch.int8 | torch.CharTensor | torch.cuda.CharTensor
16-bit integer (signed) | torch.int16 or torch.short | torch.ShortTensor | torch.cuda.ShortTensor
32-bit integer (signed) | torch.int32 or torch.int | torch.IntTensor | torch.cuda.IntTensor
64-bit integer (signed) | torch.int64 or torch.long | torch.LongTensor | torch.cuda.LongTensor

`torch.Tensor` is a float32 data type value
`torch.LongTensor` is a int64 data type value
